File: src/frontend/folha/lib/app/data/model/folha_lancamento_detalhe_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:folha/app/data/model/model_imports.dart';

class FolhaLancamentoDetalheModel {
	int? id;
	int? idFolhaLancamentoCabecalho;
	int? idFolhaEvento;
	double? origem;
	double? provento;
	double? desconto;
	FolhaEventoModel? folhaEventoModel;

	FolhaLancamentoDetalheModel({
		this.id,
		this.idFolhaLancamentoCabecalho,
		this.idFolhaEvento,
		this.origem,
		this.provento,
		this.desconto,
		this.folhaEventoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'origem',
		'provento',
		'desconto',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Origem',
		'Provento',
		'Desconto',
	];

	FolhaLancamentoDetalheModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idFolhaLancamentoCabecalho = jsonData['idFolhaLancamentoCabecalho'];
		idFolhaEvento = jsonData['idFolhaEvento'];
		origem = jsonData['origem']?.toDouble();
		provento = jsonData['provento']?.toDouble();
		desconto = jsonData['desconto']?.toDouble();
		folhaEventoModel = jsonData['folhaEventoModel'] == null ? FolhaEventoModel() : FolhaEventoModel.fromJson(jsonData['folhaEventoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idFolhaLancamentoCabecalho'] = idFolhaLancamentoCabecalho != 0 ? idFolhaLancamentoCabecalho : null;
		jsonData['idFolhaEvento'] = idFolhaEvento != 0 ? idFolhaEvento : null;
		jsonData['origem'] = origem;
		jsonData['provento'] = provento;
		jsonData['desconto'] = desconto;
		jsonData['folhaEventoModel'] = folhaEventoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idFolhaLancamentoCabecalho = plutoRow.cells['idFolhaLancamentoCabecalho']?.value;
		idFolhaEvento = plutoRow.cells['idFolhaEvento']?.value;
		origem = plutoRow.cells['origem']?.value?.toDouble();
		provento = plutoRow.cells['provento']?.value?.toDouble();
		desconto = plutoRow.cells['desconto']?.value?.toDouble();
		folhaEventoModel = FolhaEventoModel();
		folhaEventoModel?.nome = plutoRow.cells['folhaEventoModel']?.value;
	}	

	FolhaLancamentoDetalheModel clone() {
		return FolhaLancamentoDetalheModel(
			id: id,
			idFolhaLancamentoCabecalho: idFolhaLancamentoCabecalho,
			idFolhaEvento: idFolhaEvento,
			origem: origem,
			provento: provento,
			desconto: desconto,
		);			
	}

	
}

File: src/backend/folha/src/model/folha_lancamento_detalhe_model.py
from src import db
from src.model.folha_evento_model import FolhaEventoModel


class FolhaLancamentoDetalheModel(db.Model):
    __tablename__ = 'folha_lancamento_detalhe'

    id = db.Column(db.Integer, primary_key=True)
    origem = db.Column(db.Float)
    provento = db.Column(db.Float)
    desconto = db.Column(db.Float)
    id_folha_lancamento_cabecalho = db.Column(db.Integer, db.ForeignKey('folha_lancamento_cabecalho.id'))
    id_folha_evento = db.Column(db.Integer, db.ForeignKey('folha_evento.id'))

    folha_evento_model = db.relationship('FolhaEventoModel', foreign_keys=[id_folha_evento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_folha_lancamento_cabecalho = data.get('idFolhaLancamentoCabecalho')
        self.id_folha_evento = data.get('idFolhaEvento')
        self.origem = data.get('origem')
        self.provento = data.get('provento')
        self.desconto = data.get('desconto')

    def serialize(self):
        return {
            'id': self.id,
            'idFolhaLancamentoCabecalho': self.id_folha_lancamento_cabecalho,
            'idFolhaEvento': self.id_folha_evento,
            'origem': self.origem,
            'provento': self.provento,
            'desconto': self.desconto,
            'folhaEventoModel': self.folha_evento_model.serialize() if self.folha_evento_model else None,
        }